File: 14-LongestCommonPrefix/14-LongestCommonPrefix.py
# Last updated: 1/8/2026, 5:29:22 p.m.
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        current = self.root

        for char in word:
            if char not in current.children:
                current.children[char] = TrieNode()

            current = current.children[char]
        current.is_end_of_word = True

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        trie = Trie()
        for string in strs:
            trie.insert(string)

        current = trie.root
        prefix = []

        while len(current.children) == 1 and not current.is_end_of_word:
            char = list(current.children.keys())[0]
            prefix.append(char)
            current = current.children[char]

        return ''.join(prefix)

File: 14-LongestCommonPrefix/14-LongestCommonPrefix.c
// Last updated: 1/8/2026, 5:29:26 p.m.
char* longestCommonPrefix(char** strs, int strsSize) {
    char i = 1, j = 0, *ans = (char*)malloc(sizeof(char)*201);

    while(strs[0][j]){
        while(i < strsSize && strs[0][j] == strs[i][j]) i++;
        if(i != strsSize) break;

        ans[j] = strs[0][j];
        i = 1; j++;
    }
    
    ans[j] = '\0';
    return ans;
}